src/main/java/jenkins/plugins/git/GitSCMFileSystem.java

```diff
@@ -286,40 +289,80 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
+                                            @CheckForNull String refSpec,
+                                            @CheckForNull EnvVars env,
+                                            @CheckForNull String remoteName) {
+
                 String branchSpecExpandedName = branchSpec.getName();
                 if (env != null) {
                     branchSpecExpandedName = env.expand(branchSpecExpandedName);
                 }
+                String refspecExpandedName = refSpec;
+                if (env != null && refspecExpandedName != null) {
+                    refspecExpandedName = env.expand(refspecExpandedName);
+                }
 
+                // default to a branch (refs/heads)
                 String prefix = Constants.R_HEADS;
+                // check for a tag
                 if (branchSpecExpandedName.startsWith(Constants.R_TAGS)) {
                     prefix = Constants.R_TAGS;
+                } else {
+                    // check for FETCH_HEAD
+                    if (branchSpecExpandedName.equals(Constants.FETCH_HEAD) && refspecExpandedName != null &&
+                            !refspecExpandedName.equals("")) {
+                        prefix = null;
+                    } else {
+                        // check for commit-id
+                        final String regex = "^[a-fA-F0-9]{40}$";
+                        final Pattern pattern = Pattern.compile(regex, Pattern.MULTILINE);
+                        final Matcher matcher = pattern.matcher(branchSpecExpandedName);
+
+                        if (matcher.find()) {
+                            // commit-id
+                            prefix = null;
+                            rev = new AbstractGitSCMSource.SCMRevisionImpl(new SCMHead(branchSpecExpandedName), branchSpecExpandedName);
+                        }
+                    }
                 }
 
-                String headName;
+                String calculatedHeadName = branchSpecExpandedName;
                 if (rev != null && env != null) {
-                    headName = env.expand(rev.getHead().getName());
+                    calculatedHeadName = env.expand(rev.getHead().getName());
                 } else {
-                    if (branchSpecExpandedName.startsWith(prefix)) {
-                        headName = branchSpecExpandedName.substring(prefix.length());
+                    if (prefix != null && branchSpecExpandedName.startsWith(prefix)) {
+                        calculatedHeadName = branchSpecExpandedName.substring(prefix.length());
                     } else if (branchSpecExpandedName.startsWith("*/")) {
-                        headName = branchSpecExpandedName.substring(2);
+                        calculatedHeadName = branchSpecExpandedName.substring(2);
+                    }
+                }
+
+                if (refspecExpandedName == null || refspecExpandedName.equals("")) {
+                    if (prefix.equals(Constants.R_TAGS)) {
```

Comment:
I'd suggest adding a test with a sha1 and a `null` refspec. This `if` would need to be fixed
```suggestion
                    if (prefix != null && prefix.equals(Constants.R_TAGS)) {
```